pytorch_lightning/loops/training_epoch_loop.py

```diff
@@ -41,17 +41,17 @@ def __init__(self, min_steps: int, max_steps: int):
         self.iteration_count: int = 0
         # the current split index when the batch gets split into chunks in truncated backprop through time
         self.split_idx: Optional[int] = None
-
-        self._dataloader_idx: Optional[int] = None
-        self._should_stop: bool = False
-
-        self.is_last_batch: Optional[bool] = None
+        # the number of batches seen this run, updates immediately after batch_loop.run()
         self.batches_seen: int = 0
-        self.warning_cache: WarningCache = WarningCache()
-        self.epoch_output: Optional[List[List[STEP_OUTPUT]]] = None
 
         self.batch_loop: Optional[TrainingBatchLoop] = None
 
+        self._dataloader_idx: Optional[int] = None
+        self._should_stop: bool = False
+        self._is_last_batch: Optional[bool] = None
```

Comment:
Agent underscore will have to leave this one alone

```python
$ grep -irn "is_last_batch" docs
docs/source/common/optimizers.rst:230:        if self.trainer.is_last_batch and (self.trainer.current_epoch + 1) % n == 0:
docs/source/guides/speed.rst:386:            is_last_batch_to_accumulate = \
docs/source/guides/speed.rst:387:                (batch_idx + 1) % 2 == 0 or self.trainer.is_last_batch
```

Reply:
oh, we don't actually have a is_last_batch on the Trainer properties hmm. Either have to re-introduce it or defer to `trainer.fit_loop.training_loop.is_last_batch` in docs

Reply:
I would reintroduce it